src/Core/Parser/SyntaxTree/Expression/TernaryExpressionNode.php

```diff
@@ -123,23 +122,27 @@ public function compile(TemplateCompiler $templateCompiler)
         $parts = array_map([__CLASS__, 'trimPart'], $parts);
         list($check, $then, $else) = $parts;
 
+        if ($then === '') {
+            $then = $check[0] === '!' ? $else : $check;
+        }
+
         $matchesVariable = $templateCompiler->variableName('array');
         $initializationPhpCode = '// Rendering TernaryExpression node' . chr(10);
         $initializationPhpCode .= sprintf('%s = %s;', $matchesVariable, var_export($this->getMatches(), true)) . chr(10);
 
-        $parser = new BooleanParser();
-        $compiledExpression = $parser->compile($check);
         $functionName = $templateCompiler->variableName('ternaryExpression');
         $initializationPhpCode .= sprintf(
             '%s = function($context, $renderingContext) {
-				if (%s::convertToBoolean(' . $compiledExpression . ', $renderingContext) === TRUE) {
-					return %s::getTemplateVariableOrValueItself(%s, $renderingContext);
-				} else {
-					return %s::getTemplateVariableOrValueItself(%s, $renderingContext);
-				}
+                $check = %s;
+                $parser = new \TYPO3Fluid\Fluid\Core\Parser\BooleanParser();
+                $checkResult = $parser->evaluate($check, $context);
```

Comment:
This is the main change: We're not using ->compile() anymore for the time being, but simply parse $check again:

It seems the code returned by ->compile() has at least some quoting issues.

We switch to full evaluation here for the time being to fix the cache-related bugs first.

Looking at ->compile() and maybe using it here again to optimize evaluation is another task, but fixing the main issues of cached-behavior has priority for now.